Make file descriptors from stdlib non-inheritable by default

alaviss opened the pull request:
Fixes #6602. Please refer to the linked issue for the rationale.

This is still pretty much a WIP. The final goal here is to make all file descriptors created from stdlib's high-level abstractions non-inheritable by default.

alaviss committed:
io: make file descriptors non-inheritable by default

This prevents file descriptors/handles leakage to child processes
that might cause issues like running out of file descriptors, or potential
security issues like leaking a file descriptor to a restricted file.

While this breaks backward compatibility, I'm rather certain that not
many programs (if any) actually make use of this implementation detail.
A new API `setInheritable` is provided for the few that actually want to
use this functionality.

Varriount commented:
@alaviss and others: This might be useful - https://www.python.org/dev/peps/pep-0446/

Araq commented:
I'm not a fan of fixing UNIX bugs in Nim, but I won't fight it.

alaviss commented:
> I'm not a fan of fixing UNIX bugs in Nim, but I won't fight it.

Windows leak handles too, unless you manually use Win32 API.

Araq commented:
> Windows leak handles too, unless you manually use Win32 API.

I always wanted to use the Win32 API directly for IO indeed.

alaviss committed:
io: disable inheritance at file creation time for supported platforms

Some platforms provide extension to fopen-family of functions to allow
for disabling descriptor inheritance atomically during File creation.
This guards against possible leaks when a child process is spawned
before we managed to disable the file descriptor inheritance
(ie. in a multi-threaded program).

alaviss committed:
net, nativesockets: make sockets non inheritable by default

With this commit, sockets will no longer leak to child processes when
you don't want it to. Should solves a lot of "address in use" that might
occur when your server has just restarted.

All APIs that create sockets in these modules now expose a `inheritable`
flag that allow users to toggle inheritance for the resulting sockets.
An implementation of `setInheritance()` is also provided for SocketHandle.

While atomically disabling inheritance at creation time is supported on
Windows, it's only implemented by native winsock2, which is too much for
now. This support can be implemented in a future patch.

disruptek commented:
But isn't this a heavy-handed change for the language to impose on the user?  How does the user get around it?

alaviss commented:
> But isn't this a heavy-handed change for the language to impose on the user? How does the user get around it?

By marking file descriptors that you want to pass down with `setInheritable()`.

And I don't even think most users will ever notice these changes. Maybe they will find out that their web server no longer error out due to "address in use" when it restarts (see #6602), but for the few that actually use this functionality, chances are that they know how to do it correctly.

disruptek commented:
That makes sense; thanks for explaining it.

alaviss committed:
posix: add F_DUPFD_CLOEXEC

This command duplicates file descriptor with close-on-exec flag set.

Defined in POSIX.1-2008.

alaviss committed:
ioselectors_kqueue: don't leak file descriptors

File descriptors internally used by ioselectors on BSD/OSX are now
shielded from leakage.

alaviss committed:
posix: add O_CLOEXEC

This flag allows file descriptors to be open() with close-on-exec flag
set atomically.

This flag is specified in POSIX.1-2008

alaviss committed:
tfdleak: test for selectors leakage

Also simplified the test by using handle-type agnostic APIs to test for
validity.

alaviss committed:
ioselectors_epoll: mark all fd created close-on-exec

File descriptors from ioselectors should no longer leaks on Linux.

alaviss commented:
That should be about all of them. Please let me know if I missed any.

alaviss commented:
Reopening for Windows CI which failed for unrelated reasons.

alaviss committed:
tfdleak: don't check for selector leakage on Windows

The getFd proc for ioselectors_select returns a hardcoded -1

Varriount commented:
It's unfortunate there's no way to enable a test at compile time to list or detect inherited file descriptors

Araq commented on lib/system/io.nim:
```diff
@@ -488,6 +521,14 @@ const
     #"rt", "wt", "w+t", "r+t", "at"
     # we always use binary here as for Nim the OS line ending
     # should not be translated.
+  NoInheritFlag =
```
Move this above ``FormatOpen`` and do the concats at compile-time in the ``FormatOpen`` array.

Araq commented on lib/system/io.nim:
```diff
@@ -527,17 +568,24 @@ proc open*(f: var File, filename: string,
-  var p = fopen(filename, FormatOpen[mode])
+  ##
+  ## The file handle associated with the resulting ``File`` is not inheritable.
+  var p = fopen(filename, FormatOpen[mode] & NoInheritFlag)
```
And then use no string concats here.

Araq commented:
> Please let me know if I missed any.

Ha, good one. I don't know either.

alaviss committed:
io: add NoInheritFlag at compile time

alaviss committed:
io: add support for ioctl-based close-on-exec

This allows for the flag to be set/unset in one syscall. While the
performance gains might be negliable, we have one less failure point
to deal with.

alaviss committed:
tfdleak: add a test for setInheritable

Araq commented:
There has been concerns about how this makes multi-process servers much harder to write than before. Can we add a build option like ``-d:nimInheritHandles`` for these cases and also for backwards compatibility?

alaviss commented:
> There has been concerns about how this makes multi-process servers much harder to write than before. Can we add a build option like `-d:nimInheritHandles` for these cases and also for backwards compatibility?

I'll look into it, though I'd say that this should be done for backward compatibility only. Newer code should be updated to explicitly declare which sockets should be passed down (the most important part in writing multi-process servers imo). Ideally some discussions on what would be of concern for multi-process servers is better to provide adequate support for those cases.

alaviss committed:
stdlib: add nimInheritHandles to restore old behaviors

alaviss committed:
memfiles: make file handle not inheritable by default for posix

alaviss committed:
io: setInheritable now operates on OS file handle

On Windows, the native handle is the only thing that's inheritable, thus
we can assume that users of this function will already have the handle
available to them. This also allows users to pass down file descriptors
from memfiles on Windows with ease, should that be desired.

With this, nativesockets.setInheritable can be made much simpler.

Araq commented on changelog.md:
```diff
@@ -21,7 +21,12 @@
+  to allow marking explicitly which file handles should be inherited.
+  For a transistion period, define `nimInheritHandles` to enable file handle
+  inheritance by default. This flag does **not** affect the `selectors` module
+  due to the differing semantics between operating systems.
```
But later on you write this

> File handles created via `system`, `nativesockets`, `net`, and `selectors` modules are no longer inheritable by default.

So which is it?

alaviss replied:
This one is more of an overview, but you're right, I should clarify it.

Araq commented on lib/pure/nativesockets.nim:
```diff
@@ -185,19 +185,52 @@ proc toSockType*(protocol: Protocol): SockType =
+
+when declared(setInheritable) or defined(nimdoc):
+  proc setInheritable*(s: SocketHandle, inheritable: bool): bool {.inline.} =
+    ## Set whether a socket is inheritable by child processes.
```
What's the return value? Also, this shouldn't have a return value.

alaviss replied:
The return value is whether `setInheritable` succeed. It has one because the `io.setInheritable` have one, for consistency purposes.

Do you think I should `raise` instead? Though most of the APIs here use error signaling rather than exceptions.

Araq replied:
Yeah, ok, so document it well.

alaviss committed:
changelog: clarify

alaviss committed:
nativesockets: document setInheritable return value

alaviss committed:
Merge branch 'devel' into fd-no-leak

alaviss committed:
Merge branch 'devel' into fd-no-leak

kubo39 commented on lib/pure/ioselects/ioselectors_epoll.nim:
```diff
@@ -81,7 +81,7 @@ proc newSelector*[T](): Selector[T] =
   const numFD = 1024
 
-  var epollFD = epoll_create(MAX_EPOLL_EVENTS)
+  var epollFD = epoll_create1(O_CLOEXEC)
```
epoll_create1 was added to the Linux in version 2.6.27. Is it ok for Nim's minimal supported Linux version?
http://man7.org/linux/man-pages/man2/epoll_create.2.html

alaviss replied:
It's released over a decade ago, I don't see why not :)

Araq replied:
Hmm, but isn't Linux 2.4 still widely used for embedded development?

kubo39 replied:
We don't have to care about that, since ioselectors_epoll uses eventfd(2) which is added in Linux 2.6.22.

kubo39 commented on lib/pure/ioselects/ioselectors_epoll.nim:
```diff
@@ -110,7 +110,7 @@ proc close*[T](s: Selector[T]) =
 
 proc newSelectEvent*(): SelectEvent =
-  let fdci = eventfd(0, 0)
+  let fdci = eventfd(0, O_CLOEXEC)
```
evenfd's man says:
> In Linux up to version 2.6.26, the flags argument is unused, and must be specified as zero.
https://linux.die.net/man/2/eventfd

Also, we can set `O_NONBLOCK` too here, and remove setNonBlocking.

alaviss replied:
That's out of scope for this PR, sorry.

I'd like to focus on one change at a time, that way we can better trace any failure back to the source (esp. since this one is really big). But I do plan to get rid of `setNonBlocking()`.

kubo39 commented on lib/pure/ioselects/ioselectors_epoll.nim:
```diff
@@ -269,7 +269,7 @@ proc registerTimer*[T](s: Selector[T], timeout: int, oneshot: bool,
     newTs: Itimerspec
     oldTs: Itimerspec
-  let fdi = timerfd_create(CLOCK_MONOTONIC, 0).int
+  let fdi = timerfd_create(CLOCK_MONOTONIC, O_CLOEXEC).int
```
Same as above.

kubo39 commented on lib/pure/ioselects/ioselectors_epoll.nim:
```diff
@@ -314,7 +314,7 @@ when not defined(android):
     blockSignals(nmask, omask)
 
-    let fdi = signalfd(-1, nmask, 0).int
+    let fdi = signalfd(-1, nmask, O_CLOEXEC).int
```
Same as above.

kubo39 commented on lib/pure/ioselects/ioselectors_epoll.nim:
```diff
@@ -341,7 +341,7 @@ when not defined(android):
     blockSignals(nmask, omask)
 
-    let fdi = signalfd(-1, nmask, 0).int
+    let fdi = signalfd(-1, nmask, O_CLOEXEC).int
```
Same as above.

alaviss committed:
Merge branch 'devel' into fd-no-leak

kubo39 commented:
I find one more case: `mkstemp` in posix_utils.nim, and could be replaced with `mkostemp` in some OSes.

alaviss committed:
posix_utils: atomically disable fd inheritance for mkstemp

alaviss commented:
@Araq It's ready for review.

alaviss committed:
Merge branch 'devel' into fd-no-leak
